### PR TITLE
Add tests for knocking

### PR DESCRIPTION
**Pull request by anoadragon453**
This PR adds tests for knocking. They are currently hidden behind a build tag. To run the tests, you'll need to add `-tags msc2403` to your command line args to include these tests. The build flag, as well as all of the `xyz.amorgan.knock` references should be remove/modified after MSC2403 makes it into a spec version.

* MSC: https://github.com/matrix-org/matrix-doc/pull/2403
* Synapse implementation: https://github.com/matrix-org/synapse/pull/6739

### CODE REVIEW

File: internal/client/client.go
Comment (Kegsay) [inline internal/client/client.go:107]: I don't think knocking is _that_ common that it should pollute the `CSAPI` client. Move this to the test file please.

Comment (Kegsay) [reply internal/client/client.go:107]: If you require `syncUntil` logic, make that public instead.

File: tests/knocking_test.go
Comment (Kegsay) [inline tests/knocking_test.go:1]: Prefer the MSC number please, as feature names are ambiguous (particularly if there are multiple competing MSCs for the same feature.

Comment (Kegsay) [reply tests/knocking_test.go:1]: Also please add the MSC number to the filename, e.g like https://github.com/matrix-org/complement/blob/kegan/msc2836/tests/msc2836_test.go

Comment (anoadragon453) [reply tests/knocking_test.go]: Good point, much more scalable indeed.

File: tests/msc2403_test.go
Comment (clokep) [inline tests/msc2403_test.go:136]: Isn't it not even defined in the spec?

File: tests/msc2403_test.go
Comment (clokep) [inline tests/msc2403_test.go:194]: Is this a copy and paste error? Shouldn't the user be being banned, not invited?

File: tests/msc2403_test.go
Comment (clokep) [inline tests/msc2403_test.go:175]: I think a knock after this isn't allowed and should 403?

File: tests/msc2403_test.go
Comment (anoadragon453) [reply tests/msc2403_test.go:136]: *Rejecting* a knock over federation is currently not possible, but rescinding a knock over federation can work (implementations just need to figure out which homeserver to send the leave event through).

https://github.com/matrix-org/matrix-doc/pull/2403/files#diff-9dcacf9adb620c5346eba2f70f71bd50ed4f85a5e2844ac130e112165ce35c06R118-R119

File: tests/msc2403_test.go
Comment (anoadragon453) [reply tests/msc2403_test.go:194]: Err, yes. This is why it's handy to have different error codes per situation :(

File: tests/msc2403_test.go
Comment (anoadragon453) [reply tests/msc2403_test.go:175]: True! I'll stick in a test for that, thanks!

**Comment by anoadragon453** [comment]
@Kegsay heads up I'm looking for a review from your side as well just to take a look over whether these tests were written in a way you think works with Complement :)

File: tests/msc2403_test.go
Comment (clokep) [inline tests/msc2403_test.go:224]: Is this really testing what we think it is or is it testing that the user is already in the room? (Or does banning also kick the user?)

File: tests/msc2403_test.go
Comment (clokep) [inline tests/msc2403_test.go:44]: What do we do when this is standardized? Do we just update these? Do we somehow let a server specify that they support the MSC vs. the standardized version?

**Review by clokep — APPROVED** [review]
I think the logic looks sound! I didn't closely look at the changes to the blueprint/client.

File: tests/msc2403_test.go
Comment (anoadragon453) [reply tests/msc2403_test.go:224]: Banning will kick a user, yes. It updates their membership to `ban`, removing them from the room.

File: tests/msc2403_test.go
Comment (anoadragon453) [reply tests/msc2403_test.go:44]: Once the MSC is merged into a released version of the spec, Complement should be updated to support that release. Part of that update will be stabilizing knocks.

The necessary steps for stabilizing will be:

* Updating instances of `xyz.amorgan.knock` to `knock`.
* Update endpoints from `/_matrix/client/unstable/xyz.amorgan.knock/...` to `/_matrix/client/rX/...`.
* Removing the `// +build ...` line at the top so the tests are run by default (they're currently opt in).
* Updating the name of the test file to `knock_test.go` or somesuch.
* Removing the conditional on rescinding knocks over federation. Synapse will need to support that if the spec says it's possible.

It might be a good idea to put the above list into the top of the file actually.

File: tests/msc2403_test.go
Comment (Kegsay) [inline tests/msc2403_test.go:263]: You can use `must.NotError(t, msg, err)`

File: tests/msc2403_test.go
Comment (Kegsay) [inline tests/msc2403_test.go:74]: Don't do this, use `CSAPI.SendEventSynced(t, roomID, b.Event)` to send state events into the room. This will wait for the response to come down `/sync` which is critical to reduce racey tests.

File: tests/msc2403_test.go
Comment (Kegsay) [inline tests/msc2403_test.go:118]: I would make `const constKnock = "xyz.amorgan.knock"` and use that, which will also help for reducing the number of changes when it lands.

File: tests/msc2403_test.go
Comment (Kegsay) [inline tests/msc2403_test.go:146]: You should probably wait for the leave to come down sync.

File: tests/msc2403_test.go
Comment (Kegsay) [inline tests/msc2403_test.go:164]: Wait for it to come down sync before knocking again.

File: tests/msc2403_test.go
Comment (Kegsay) [inline tests/msc2403_test.go:228]: Wait for it to come down sync.

**Review by Kegsay — CHANGES_REQUESTED** [review]
Few things but broadly LGTM

File: tests/msc2403_test.go
Comment (clokep) [reply tests/msc2403_test.go:44]: I guess my broader question is -- does complement have a way for homeservers to say what version of the spec they support? Not expecting it here necessarily, just bringing it up for conversation!

And if it does, does it make sense for a homeserver to say it supports 0.6 + msc2403 even after 0.7 is released?

Comment (anoadragon453) [reply tests/msc2403_test.go:44]: @Kegsay I'd be interested in your answer here in case it differs from mine.

I don't think it does, and in doing so I think Complement would need to support multiple spec versions, which I don't think it's planning to. Mostly because new spec versions often require modifying existing tests instead of making a separate set of new ones.

Selecting a range of tests is done via build tags, thus I think this choice would be made in your individual CI script rather than in a config file in the server/complement (at least currently!).

Comment (clokep) [reply tests/msc2403_test.go:44]: Alright. I don't think we need to solve this here, was just a thought I had that!

**Comment by anoadragon453** [comment]
@Kegsay quite a few chunky changes since your last review, all addressing your feedback. Details are in commit messages, but overall it was mostly a lot of code for syncing, part of which required some slight changes to client.go.

**Review by Kegsay — APPROVED** [review]
We'll probably want to factor out the leave synced stuff so we don't need to hit initial sync so much but for now this will do. LGTM!

**Pull request merged**
